### PR TITLE
feat: Add caching

### PR DESCRIPTION
**Pull request by nadeesha**
Add caching to differential functions.

```ts
// src/services/product.ts

const getProduct = async (productId: string) => {
  const product = await getProductFromDatabase(productId);
  return product;
};

export const productService = d.service({
  name: "product",
  functions: {
    getProduct: cached(getProduct, 10000), // cache results for 10 seconds
  },
});

// src/client.ts

const productClient = d.client<typeof productService>("product");

const productId = "123";

const product = await productClient.getProduct(productId, {
  $cacheKey: productId,
});

// if you call the function again with the same cache key, previous result will be returned

const product2 = await productClient.getProduct(productId, {
  $cacheKey: productId,
});

assert.deepEqual(product === product2);
```

### CODE REVIEW

File: control-plane/src/modules/service-definitions.ts [hunk 1]
```diff
@@ -1,7 +1,13 @@
 import { and, eq } from "drizzle-orm";
+import NodeCache from "node-cache";
+import { z } from "zod";
 import * as data from "./data";
 import { backgrounded } from "./util";
-import { z } from "zod";
+
+const cache = new NodeCache({
```
Comment (johnjcsmith) [inline control-plane/src/modules/service-definitions.ts:7]: I wonder if we could cache function results in `ts-core` also based on the cache key, so they don't even hit the control-plane?

Comment (nadeesha) [reply control-plane/src/modules/service-definitions.ts:7]: Yeah definitely. I think this is doable. But to do that we have to ferry the service definition to the client. Or, just encode the cache config in the client call.

**Review by johnjcsmith — APPROVED** [review]
🙌

File: control-plane/src/modules/jobs.ts [hunk 2]
```diff
@@ -2,44 +2,164 @@ import { and, eq, sql } from "drizzle-orm";
...
 import {
   ServiceDefinition,
+  getServiceDefinitions,
   storeServiceDefinitionBG,
 } from "./service-definitions";
+import { backgrounded } from "./util";
 
-export const createJob = async ({
+const createJobStrategies = {
+  idempotece: async ({
```
Comment (isaacsu) [inline control-plane/src/modules/jobs.ts:14]: speling
```suggestion
  idempotence: async ({
```

Comment (nadeesha) [reply control-plane/src/modules/jobs.ts]: Thank you!

**Pull request merged**